openmdao/drivers/genetic_algorithm_driver.py

```diff
@@ -0,0 +1,538 @@
+                fitness[max_index] = min_fit
+
+            # Find best performing point in this generation.
+            min_fit = np.min(fitness)
+            min_index = np.argmin(fitness)
+            min_gen = old_gen[min_index]
+            min_x = x_pop[min_index]
+
+            if min_fit < fopt:
+                fopt = min_fit
+                xopt = min_x
+
+            # Evolve new generation.
+            new_gen = self.tournament(old_gen, fitness)
+            new_gen = self.crossover(new_gen, Pc)
+            new_gen = self.mutate(new_gen, Pm)
+
+        return xopt, fopt, nfit
+
+    def tournament(self, old_gen, fitness):
+        """
+        Apply tournament selection and keep the best points.
+
+        Parameters
+        ----------
+        old_gen : ndarray
+            Points in current generation
+
+        fitness : ndarray
+            Objective value of each point.
+
+        Returns
+        -------
+        ndarray
+            New generation with best points.
+        """
+        new_gen = []
+        idx = np.array(range(0, self.npop - 1, 2))
+        for j in range(2):
+            old_gen, i_shuffled = self.shuffle(old_gen)
+            fitness = fitness[i_shuffled]
+
+            # Each point competes with its neighbor; save the best.
+            i_min = np.argmin(np.array([[fitness[idx]], [fitness[idx + 1]]]), axis=0)
+            selected = i_min + idx
+            new_gen.append(old_gen[selected])
+
+        return np.concatenate(np.array(new_gen), axis=1).reshape(old_gen.shape)
+
+    def crossover(self, old_gen, Pc):
+        """
+        Apply crossover to the current generation.
+
+        Crossover flips two adjacent genes.
+
+        Parameters
+        ----------
+        old_gen : ndarray
+            Points in current generation
+
+        Pc : float
+            Probability of crossover.
+
+        Returns
+        -------
+        ndarray
+            Current generation with crossovers applied.
+        """
+        new_gen = copy.deepcopy(old_gen)
+        num_sites = self.npop // 2
+        sites = np.random.rand(num_sites, self.lchrom)
+        idx, idy = np.where(sites < Pc)
+        for ii, jj in zip(idx, idy):
+            i = 2 * ii
+            j = i + 1
+            new_gen[i][jj] = old_gen[j][jj]
+            new_gen[j][jj] = old_gen[i][jj]
+        return new_gen
+
+    def mutate(self, current_gen, Pm):
+        """
+        Apply mutations to the current generation.
+
+        A mutation flips the state of the gene from 0 to 1 or 1 to 0.
+
+        Parameters
+        ----------
+        current_gen : ndarray
+            Points in current generation
+
+        Pm : float
+            Probability of mutation.
+
+        Returns
+        -------
+        ndarray
+            Current generation with mutations applied.
+        """
+        temp = np.random.rand(self.npop, self.lchrom)
+        idx, idy = np.where(temp < Pm)
+        for ii, jj in zip(idx, idy):
```

Comment:
could this loop just be replaced with `current_gen[idx, idy] = 1 - current_gen[idx, idy]`?

Reply:
Looks like it works.